domain/status/state/state.go

```diff
@@ -242,6 +242,126 @@ ON CONFLICT(application_uuid) DO UPDATE SET
+	empty := status.StatusInfo[status.RelationStatusType]{}
+	id := relationUUID{
+		RelationUUID: uuid,
+	}
+	var sts relationStatus
+
+	stmt, err := st.Prepare(`
+SELECT &relationStatus.*
+FROM   relation_status
+WHERE  relation_uuid = $relationUUID.relation_uuid
+`, id, sts)
+	if err != nil {
+		return empty, errors.Capture(err)
+	}
+
+	err = tx.Query(ctx, stmt, id).Get(&sts)
+	if errors.Is(err, sqlair.ErrNoRows) {
+		return empty, statuserrors.RelationNotFound
+	} else if err != nil && !errors.Is(err, sqlair.ErrNoRows) {
+		return empty, errors.Capture(err)
+	}
+
+	statusType, err := status.DecodeRelationStatus(sts.StatusID)
+	if err != nil {
+		return empty, errors.Capture(err)
+	}
+	return status.StatusInfo[status.RelationStatusType]{
+		Status:  statusType,
+		Message: sts.Reason,
+		Since:   sts.Since,
+	}, nil
+}
+
+// SetRelationStatus saves the given relation status, overwriting any current
+// status data. It can return the following errors:
+//   - [statuserrors.RelationNotFound] if the relation doesn't exist.
+//   - [statuserrors.RelationStatusTransitionNotValid] if the current relation
+//     status cannot transition to the new relation status. the relation does
+//     not exist.
+func (st *State) SetRelationStatus(
+	ctx context.Context,
+	relationUUID corerelation.UUID,
+	sts status.StatusInfo[status.RelationStatusType],
+) error {
+	db, err := st.DB()
+	if err != nil {
+		return errors.Capture(err)
+	}
+
+	err = db.Txn(ctx, func(ctx context.Context, tx *sqlair.TX) error {
+		// Get current status.
+		currentStatus, err := st.getRelationStatus(ctx, tx, relationUUID)
+		if err != nil {
+			return errors.Errorf("getting current relation status: %w", err)
+		}
+
+		// Check we can transition from current status to the new status.
+		err = status.RelationStatusTransitionValid(currentStatus, sts)
+		if err != nil {
+			return errors.Capture(err)
+		}
+
+		// If transitioning from Suspending to Suspended, retain any existing
+		// message so that any previous reason for suspending is retained.
+		if sts.Message == "" &&
+			currentStatus.Status == status.RelationStatusTypeSuspending &&
+			sts.Status == status.RelationStatusTypeSuspended {
+			sts.Message = currentStatus.Message
+		}
+		return st.upsertRelationStatus(ctx, tx, relationUUID, sts)
+	})
+	if err != nil {
+		return errors.Errorf("updating relation status for %q: %w", relationUUID, err)
+	}
+	return nil
+}
+
+func (st *State) upsertRelationStatus(
+	ctx context.Context,
+	tx *sqlair.TX,
+	relationUUID corerelation.UUID,
+	sts status.StatusInfo[status.RelationStatusType],
+) error {
+	statusID, err := status.EncodeRelationStatus(sts.Status)
+	if err != nil {
+		return errors.Capture(err)
+	}
+
+	statusInfo := relationStatus{
+		RelationUUID: relationUUID,
+		StatusID:     statusID,
+		Reason:       sts.Message,
+		Since:        sts.Since,
+	}
+	stmt, err := st.Prepare(`
+INSERT INTO relation_status (*) 
+VALUES ($relationStatus.*)
+ON CONFLICT(relation_uuid) DO UPDATE SET
```

Comment:
question: why are we doing an upset here? The relation status row is created when during AddRelation.

Reply:
Good point, I think when I wrote it there was still some confusion about whether it was. Changed to an updated.